File: src/Z.Reflection/GetDeclaration/ParameterInfo.GetDeclaration.cs
// Description: C# Extension Methods Library to enhances the .NET Framework by adding hundreds of new methods. It drastically increases developers productivity and code readability. Support C# and VB.NET
// Website & Documentation: https://github.com/zzzprojects/Z.ExtensionMethods
// Forum: https://github.com/zzzprojects/Z.ExtensionMethods/issues
// License: https://github.com/zzzprojects/Z.ExtensionMethods/blob/master/LICENSE
// More projects: http://www.zzzprojects.com/
// Copyright © ZZZ Projects Inc. 2014 - 2016. All rights reserved.
using System;
using System.Collections.Generic;
using System.Reflection;
using System.Runtime.CompilerServices;
using System.Text;

public static partial class Extensions
{
    /// <summary>A ParameterInfo extension method that gets a declaraction.</summary>
    /// <param name="this">The @this to act on.</param>
    /// <returns>The declaraction.</returns>
    public static string GetDeclaraction(this ParameterInfo @this)
    {
        var sb = new StringBuilder();

        @this.GetDeclaraction(sb);
        return sb.ToString();
    }

    internal static void GetDeclaraction(this ParameterInfo @this, StringBuilder sb)
    {
        // retval attribute
        var attributes = new List<string>();

        string typeName;
        Type elementType = @this.ParameterType.GetElementType();

        if (elementType != null)
        {
            typeName = @this.ParameterType.Name.Replace(elementType.Name, elementType.GetShortDeclaraction());
        }
        else
        {
            typeName = @this.ParameterType.GetShortDeclaraction();
        }


        if (Attribute.IsDefined(@this, typeof (ParamArrayAttribute)))
        {
            sb.Append("params ");
        }
        else if (@this.Position == 0 && @this.Member.IsDefined(typeof (ExtensionAttribute)))
        {
            sb.Append("this ");
        }

        if (@this.IsIn)
        {
            attributes.Add("In");
        }
        if (@this.IsOut)
        {
            if (typeName.Contains("&"))
            {
                typeName = typeName.Replace("&", "");
                sb.Append("out ");
            }
            else
            {
                attributes.Add("Out");
            }
        }
        else if (@this.ParameterType.IsByRef)
        {
            typeName = typeName.Replace("&", "");
            sb.Append("ref ");
        }
        sb.Append(typeName);

        sb.Append(" ");
        sb.Append(@this.Name);

        if (@this.IsOptional)
        {
            if (@this.DefaultValue != Missing.Value)
            {
                sb.Append(" = " + @this.DefaultValue);
            }
            else
            {
                attributes.Add("Optional");
            }
        }


        string attribute = attributes.Count > 0 ? "[" + string.Join(", ", attributes) + "] " : "";
        sb.Insert(0, attribute);
    }
}

File: src (VB.NET)/Z.ExtensionMethods.VB/Z.Reflection/GetDeclaration/ParameterInfo.GetDeclaration.vb

' Copyright (c) 2015 ZZZ Projects. All rights reserved
' Licensed under MIT License (MIT) (https://github.com/zzzprojects/Z.ExtensionMethods)
' Website: http://www.zzzprojects.com/
' Feedback / Feature Requests / Issues : http://zzzprojects.uservoice.com/forums/283927
' All ZZZ Projects products: Entity Framework Extensions / Bulk Operations / Extension Methods /Icon Library

Imports System.Collections.Generic
Imports System.Reflection
Imports System.Runtime.CompilerServices
Imports System.Text

Public Module Extensions_810

	<System.Runtime.CompilerServices.Extension> _
	Public Function GetDeclaraction(this As ParameterInfo) As String
		Dim sb = New StringBuilder()

		this.GetDeclaraction(sb)
		Return sb.ToString()
	End Function

    <System.Runtime.CompilerServices.Extension> _
    Friend Sub GetDeclaraction(this As ParameterInfo, sb As StringBuilder)
        ' retval attribute
        Dim attributes = New List(Of String)()

        Dim typeName As String
        Dim elementType As Type = this.ParameterType.GetElementType()

        If elementType IsNot Nothing Then
            typeName = this.ParameterType.Name.Replace(elementType.Name, elementType.GetShortDeclaraction())
        Else
            typeName = this.ParameterType.GetShortDeclaraction()
        End If


        If Attribute.IsDefined(this, GetType(ParamArrayAttribute)) Then
            sb.Append("params ")
        ElseIf this.Position = 0 AndAlso this.Member.IsDefined(GetType(ExtensionAttribute)) Then
            sb.Append("this ")
        End If

        If this.IsIn Then
            attributes.Add("In")
        End If
        If this.IsOut Then
            If typeName.Contains("&") Then
                typeName = typeName.Replace("&", "")
                sb.Append("out ")
            Else
                attributes.Add("Out")
            End If
        ElseIf this.ParameterType.IsByRef Then
            typeName = typeName.Replace("&", "")
            sb.Append("ref ")
        End If
        sb.Append(typeName)

        sb.Append(" ")
        sb.Append(this.Name)

        If this.IsOptional Then
            If this.DefaultValue <> Missing.Value Then
                sb.Append(" = " + this.DefaultValue)
            Else
                attributes.Add("Optional")
            End If
        End If


        Dim attribute__1 As String = If(attributes.Count > 0, "[" + String.Join(", ", attributes) + "] ", "")
        sb.Insert(0, attribute__1)
    End Sub
End Module


